cursive-core/src/logger.rs

```diff
@@ -2,14 +2,107 @@
+///
+/// ```
+/// # use cursive_core::logger::CursiveLogger;
+/// CursiveLogger::new()
+///     .with_env()
+///     .init();
+/// ```
+///
+/// Set log levels explicitly.
+///
+/// ```
+/// # use cursive_core::logger::CursiveLogger;
+/// # use log::LevelFilter;
+/// CursiveLogger::new()
+///     .with_int_filter_level(LevelFilter::Warn)
+///     .with_ext_filter_level(LevelFilter::Debug)
+///     .init();
+/// ```
+///
+/// Set log queue size.
+///
+/// ```
+/// # use cursive_core::logger::CursiveLogger;
+/// CursiveLogger::new()
+///     .with_log_size(10_000)
+///     .init();
+/// ```
+pub struct CursiveLogger {
+    // Log filter level for log messages from within cursive
+    int_filter_level: log::LevelFilter,
+    // Log filter level for log messages from sources outside of cursive
+    ext_filter_level: log::LevelFilter,
+    // Size of log queue
+    log_size: usize,
+}
+
+impl CursiveLogger {
+    /// Creates a new CursiveLogger with default log filter levels of `log::LevelFilter::Trace`.
+    /// Remember to call `init()` to install with `log` backend.
+    pub fn new() -> Self {
+        CursiveLogger {
+            int_filter_level: log::LevelFilter::Trace,
+            ext_filter_level: log::LevelFilter::Trace,
+            log_size: 1000,
+        }
+    }
+
+    /// Sets the internal log filter level.
+    pub fn with_int_filter_level(mut self, level: log::LevelFilter) -> Self {
+        self.int_filter_level = level;
+        self
+    }
+
+    /// Sets the external log filter level.
+    pub fn with_ext_filter_level(mut self, level: log::LevelFilter) -> Self {
+        self.ext_filter_level = level;
+        self
+    }
 
-static LOGGER: CursiveLogger = CursiveLogger;
+    /// Sets log filter levels based on environment variables `RUST_LOG` and `CURSIVE_LOG`.
+    /// If `RUST_LOG` is set, then both internal and external log levels are set to match.
+    /// If `CURSIVE_LOG` is set, then the internal log level is set to match with precedence over
+    /// `RUST_LOG`.
+    pub fn with_env(mut self) -> Self {
+        if let Ok(rust_log) = std::env::var("RUST_LOG") {
+            if let Ok(filter_level) = log::LevelFilter::from_str(&rust_log) {
+                self.int_filter_level = filter_level;
+                self.ext_filter_level = filter_level;
+            }
+        }
+        if let Ok(cursive_log) = std::env::var("CURSIVE_LOG") {
+            if let Ok(filter_level) = log::LevelFilter::from_str(&cursive_log) {
+                self.int_filter_level = filter_level;
+            }
+        }
+        self
+    }
+
+    /// Sets the size of the log queue
+    pub fn with_log_size(mut self, log_size: usize) -> Self {
+        self.log_size = log_size;
+        self
+    }
+
+    /// Installs the logger with log. Calling twice will panic.
+    pub fn init(self) {
+        reserve_logs(self.log_size);
+        log::set_logger(Box::leak(Box::new(self))).unwrap();
+        log::set_max_level(log::LevelFilter::Trace);
```

Comment:
We could set the maximum of (int, ext) levels to avoid logging trace if it wouldn't be seen anyway.

Reply:
Done.